docs/dag-sharding-rfc.md

```diff
@@ -0,0 +1,182 @@
+
+The golang multipart-reader exposes a reader to each part of the body in succession.  When the part's body is read completely, a reader to the next part's body can be generated by the multipart-reader.  The part meta-data (e.g. this part is a directory, that part is a file) and the data read from the part, must then be passed into the importer component to begin the next phase of the pipeline.  We can explore different methods of this transfer in pursuit of efficiency and simplicity.  One approach would be polling the part readers in the HTTP API and transferring data as it arrives over RPC, while another would be to assume the importer is on the same machine as the HTTP API and send the multi-part reader, or the readers to individual parts, over a channel.
+
+As this endpoint will be used for streaming large amounts of data it has the potential to be somewhat difficult to use as system limits could be pushed and errors could occur after annoyingly long time investments.  Some questions that may come up as useability becomes a bigger concern:
+
+* Is it possible that remote users pushing data to the endpoint could overload cluster's resources by pushing at a rate faster than processing can occur in cluster? If this is an issue how can the endpoint maintain connections in a resource-aware non-degrading way?  Will it be worthwhile to open up an endpoint running a different protocol (i.e. libp2p stream-muxer)?
+* What features can cluster provide to make the process easier?  For example if cluster could report an estimate aggregate of all of the space in its nodes' repos then users could check that a huge file will fit on the cluster before importing it.
+* What is the distribution of user preferences regarding pushing or pulling data to or into the endpoint?  Is it worth including location addressed link resolution (i.e. `ipfs-cluster-ctl add --shard http://www.awesomefilez.warez`), or file system device reading support?
+* Any others?
+
+#### Sharding existing DAGs
+An endpoint should also exist for sharding an existing DAG among an ipfs-cluster.  A user will run something like `ipfs-cluster-ctl pin --shard <cid>` and pin shards of the DAG rooted at `cid` across different nodes.  The importing step of the pipeline can be skipped in this case.  The naive implementation is to interact with the local daemon's DAG API to retrieve a stream of blocks as input to the sharding component.  We currently suspect that this is not going to work well for large DAGs that cannot fit in the node's repo.  The "Future Work" subsection: "sharding large existing DAGs" discusses some possibilities for scaling this to large DAGs.
```

Comment:
Good point, still pretty useful to take something from a huge node and distribute it among several